File: data-structures/bst.py
from __future__ import annotations
from typing import Generic, TypeVar, Union

T = TypeVar('T') 

class BST(Generic[T]):
    """A binary search tree."""

    empty = None
    def __init__(self, val: T,
                    left: Union[None, BST]=empty,
                    right: Union[None, BST]=empty):
        self.val = val
        self.left = left
        self.right = right


    def __str__(self) -> str:
        if self is BST.empty:
            return ""
        return BST.__str__(self.left) + " {0} ".format(self.val) + BST.__str__(self.right)


    def is_leaf(self) -> bool:
        return not self.left and not self.right


    def find(self, target: T) -> Union[str, T]:
        if self is BST.empty:
            return "not found"
        if self.val == target:
            return target
        elif target < self.val:
            return BST.find(self.left, target)
        return BST.find(self.right, target)


    def insert(self, node: BST) -> None:
        def insert_helper(t: BST, node: BST) -> BST:
            if t is BST.empty:
                return BST(node)
            if node < t.val:
                t.left = insert_helper(t.left, node)
            elif node > t.val:
                t.right = insert_helper(t.right, node)
            return t
        self = insert_helper(self, node)


    def delete(self, node: BST) -> None:
        def find_succ(t: BST) -> BST:
            curr_succ = t
            while t.left is not BST.empty:
                curr_succ = t.left
            return curr_succ
        if self is BST.empty:
            return
        if node < self.val:
            BST.delete(self.left, node)
        elif node > self.val:
            BST.delete(self.right, node)
        else:
            if not (self.left is BST.empty or self.right is BST.empty):
                successor = find_succ(self)
                self.val = successor.val
                successor.delete(successor.val)
            elif self.left is not BST.empty:
                self = self.left
            elif self.right is not BST.empty:
                self = self.right
            else:
                self = BST.empty


    def is_BST(self):
        if self == BST.empty or self.is_leaf():
            return True
        if self.left == BST.empty:
            return self.val < self.right.val and self.right.is_BST()
        elif self.right == BST.empty:
            return self.left.is_BST and self.val > self.left.val
        return self.left.is_BST and self.left.val < self.val < self.right.val and self.right.is_BST()


test_bst = BST(8,
                BST(3,
                    BST(1),
                    BST(6,
                        BST(4),
                        BST(7))),
                BST(10,
                    None,
                    BST(14,
                        BST(13))))
print(test_bst)

test_bst.insert(9)
print(test_bst)

test_bst.delete(9)
print(test_bst)
print(test_bst.find(9))

File: data-structures/bst.hs
data BST a = EmptyBST | BST {
    val :: a,
    left :: BST a,
    right :: BST a
} deriving (Show, Eq, Ord)

isLeaf :: (Eq a) => BST a -> Bool
isLeaf t = left t == EmptyBST && right t == EmptyBST

find :: (Ord a) => BST a -> a -> Maybe a
find EmptyBST _ = Nothing
find t x
    | val t == x = Just x
    | x < val t = find (left t) x
    | otherwise = find (right t) x

insert :: (Ord a) => BST a -> a -> BST a
insert EmptyBST x = BST x EmptyBST EmptyBST
insert t x
    | x == val t = t
    | x < val t = BST (val t) (insert (left t) x) (right t)
    | otherwise = BST (val t) (left t) (insert (right t) x)

delete :: (Ord a) => BST a -> a -> BST a
delete EmptyBST _ = EmptyBST
delete t x
    | x < val t = BST (val t) (delete (left t) x) (right t)
    | x > val t = BST (val t) (left t) (delete (right t) x)
    | otherwise = delHelper t

delHelper :: (Ord a) => BST a -> BST a
delHelper t
    | isLeaf t = EmptyBST
    | left t /= EmptyBST = left t
    | right t /= EmptyBST = right t
    | otherwise = BST (val succNode) (left t) (delete succNode (val succNode))
    where
        findSucc x
            | isLeaf x || left x == EmptyBST = x
            | otherwise = findSucc (left x)
        succNode = findSucc (right t)

isBST :: (Ord a) => BST a -> Bool
isBST EmptyBST = True
isBST t
    | isLeaf t = True
    | left t == EmptyBST = (val t < val (right t)) && isBST (right t)
    | right t == EmptyBST = isBST (left t) && (val t > val (left t))
    | otherwise = isBST (left t) && val (left t) < val t && val t < val (right t) && isBST (right t)

testBST :: BST Integer
testBST = BST 8
            (BST 3
                (BST 1 EmptyBST EmptyBST)
                (BST 6
                    (BST 4 EmptyBST EmptyBST)
                    (BST 7 EmptyBST EmptyBST)))
            (BST 10
                EmptyBST
                (BST 14
                    (BST 13 EmptyBST EmptyBST)
                    EmptyBST))

printBST :: BST a -> [a]
printBST EmptyBST = []
printBST t = printBST (left t) ++ [val t] ++ printBST (right t)
